### PR TITLE
[SYCL] Evict program cache on PI_ERROR_OUT_OF_RESOURCES

### PR DESCRIPTION
**Pull request by jzc**
This PR adds resource cleanup mechanisms when `PI_ERROR_OUT_OF_RESOURCES` occurs on `piProgramBuild` and `piProgramLink`. Currently. the whole cache is cleared, but this behavior can be extended in the future. In order to ensure thread safety of the cache clearing mechanism, the maps of the cache now hold their `BuildResult`'s in `shared_ptr`'s, which allows the cache to be cleared even if there are still threads that still hold references to a `BuildResult`.

### CODE REVIEW

File: sycl/source/detail/program_manager/program_manager.cpp [hunk 1]
```diff
@@ -135,7 +135,9 @@ KernelProgramCache::BuildResult<RetT> *
 getOrBuild(KernelProgramCache &KPCache, GetCachedBuildFT &&GetCachedBuild,
            BuildFT &&Build) {
   using BuildState = KernelProgramCache::BuildState;
+  int n_retry = 0;
 
+out:
```
Comment (steffenlarsen) [inline sycl/source/detail/program_manager/program_manager.cpp:140]: I am not a huge fan of the `goto` approach as I think it makes the control-flow hard to follow for the reader. Could we maybe instead move the body of this function to a new function and make this function do something like

```c++
for (size_t AttemptCounter = 0; AttemptCounter < MaxBuildRetry; ++AttemptCounter) {
  KernelProgramCache::BuildResult<RetT> *Result = MyNewFunction(..., /*EvictOnOutOfResources=*/true);
  if (Result ->State.load() != BuildState::BS_WillEvict)
    return Result;
}
return MyNewFunction(..., /*EvictOnOutOfResources=*/false);
```
For now `MaxBuildRetry` will just be 1 as we evict the entire cache, but in the future we can change that behavior.

The new `EvictOnOutOfResources` argument would be used instead of `n_retry == 1` further down.

**Review by steffenlarsen — COMMENTED** [review]
I like the approach, but we should probably address the race conditions before we consider merging this. Should it be converted to draft for good measure?

File: sycl/source/detail/program_manager/program_manager.cpp [hunk 2]
```diff
@@ -183,6 +188,45 @@ getOrBuild(KernelProgramCache &KPCache, GetCachedBuildFT &&GetCachedBuild,
...
+      // clearCache to not be inter-thread happen the reread.
+
+      // TODO: possible race condition - the value returned by this
+      // function is a pointer to a BR. Clearing the cache will
+      // invalidate this pointer.  In the PiProgram case, we can just
+      // the return this handle stored in the BR directly. In the
+      // PiKernel case, it's a little tricker, as other parts of the
+      // runtime (i.e. enqueueImpKernel) use a pointer to the mutex
+      // that is stored in the BR. A shared pointer should circumvent
+      // this.
```
Comment (steffenlarsen) [inline sycl/source/detail/program_manager/program_manager.cpp:225]: In a similar vein, what happens to builds that might still be ongoing but have not failed due to limited resources? Will they be evicted with the rest? If so, maybe `clearCache` should have a mode that let these stay in the cache?

File: sycl/test-e2e/KernelAndProgram/disable-caching.cpp [hunk 3]
```diff
@@ -88,6 +88,8 @@ int main() {
 
 // (Program cache releases)
 // CHECK-CACHE: piKernelRelease
-// CHECK-CACHE: piProgramRelease
 // CHECK-CACHE: piKernelRelease
+// CHECK-CACHE: piKernelRelease
+// CHECK-CACHE: piProgramRelease
+// CHECK-CACHE: piProgramRelease
```
Comment (jzc) [inline sycl/test-e2e/KernelAndProgram/disable-caching.cpp:94]: The cache is destructed in a different order now, which is why these are swapped. Also the previous version of this test only had 2 checks for release of program/kernel even though there were 3 programs/kernels that should have been created, which is why any extra check was added.

File: sycl/source/detail/kernel_program_cache.hpp [hunk 4]
```diff
@@ -203,11 +211,93 @@ class KernelProgramCache {
...
+  ///         cache. Accepts nothing. Return pointer to built entity.
+  ///
+  /// \return a pointer to cached build result, return value must not be
+  /// nullptr.
+  template <typename ExceptionT, typename GetCachedBuildFT, typename BuildFT>
+  auto getOrBuild(GetCachedBuildFT &&GetCachedBuild, BuildFT &&Build) {
+    using BuildState = KernelProgramCache::BuildState;
+    constexpr size_t MaxAttempts = 2;
+    for (size_t AttemptCounter = 0;; ++AttemptCounter) {
+      auto [BuildResult, InsertionTookPlace] = GetCachedBuild();
```
Comment (steffenlarsen) [inline sycl/source/detail/kernel_program_cache.hpp:249]: `InsertionTookPlace` seems to be unused, but we talk about insertion in the comments. Are we missing a check?

Comment (jzc) [reply sycl/source/detail/kernel_program_cache.hpp:249]: Also, using just `InsertionTookPlace` is not quite correct with how the code is structured now. In the case when the building thread fails with a non-SYCL exception, the other threads are notified to restart the build loop (via transition to `BS_Initial`). Note in this case the cache is not reset. This means that next time through the loop for all the other threads their `InsertionTookPlace` would be false, so we wouldn't be able to use that to get one of those threads to start building the `BuildResult`. (If the thread that failed initial remove the `BuildResult` from the map, then we could use `InseritonTookPlace` to get one thread to pick up the build.)

**Review by steffenlarsen — APPROVED** [review]
Just need to avoid the error, but just to avoid post-commit failure we should fix the unused variable.
